optuna/_gp/_search_space.py

```diff
@@ -0,0 +1,162 @@
+    optuna_search_space: dict[str, optuna.distributions.BaseDistribution],
+) -> tuple[SearchSpace, np.ndarray]:
+    param_type = np.zeros(len(optuna_search_space), dtype=np.int64)
+    bounds = np.zeros((len(optuna_search_space), 2), dtype=np.float64)
+    step = np.zeros(len(optuna_search_space), dtype=np.float64)
+    values = np.zeros((len(trials), len(optuna_search_space)), dtype=np.float64)
+    for i, (param, distribution) in enumerate(optuna_search_space.items()):
+        if isinstance(distribution, optuna.distributions.CategoricalDistribution):
+            param_type[i] = ParamType.CATEGORICAL
+            bounds[i, 0] = 0.0
+            bounds[i, 1] = len(distribution.choices)
+            step[i] = 1.0
+            for ti, trial in enumerate(trials):
+                values[ti, i] = distribution.to_internal_repr(trial.params[param])
+        else:
+            assert isinstance(
+                distribution,
+                (
+                    optuna.distributions.FloatDistribution,
+                    optuna.distributions.IntDistribution,
+                ),
+            )
+            if distribution.log:
+                param_type[i] = ParamType.LOG
+            else:
+                param_type[i] = ParamType.LINEAR
+            if distribution.step is None:
+                step[i] = 0.0
+            else:
+                step[i] = distribution.step
+            bounds[i, 0] = distribution.low
+            bounds[i, 1] = distribution.high
+
+            external_values = np.zeros((len(trials),))
+            for ti, trial in enumerate(trials):
+                external_values[ti] = trial.params[param]
+            values[:, i] = transform_one_param(
+                external_values, param_type[i], (bounds[i, 0], bounds[i, 1]), step[i]
+            )
+    return SearchSpace(param_type, bounds, step), values
+
+
+def get_untransformed_param(
+    optuna_search_space: dict[str, optuna.distributions.BaseDistribution],
+    transformed_param: np.ndarray,
+) -> dict[str, Any]:
+    ret = {}
+    for i, (param, distribution) in enumerate(optuna_search_space.items()):
+        if isinstance(distribution, optuna.distributions.CategoricalDistribution):
+            ret[param] = distribution.to_external_repr(transformed_param[i])
+        else:
+            assert isinstance(
+                distribution,
+                (
+                    optuna.distributions.FloatDistribution,
+                    optuna.distributions.IntDistribution,
+                ),
+            )
+            if distribution.log:
+                param_type = ParamType.LOG
+            else:
+                param_type = ParamType.LINEAR
+            if distribution.step is None:
+                step = 0.0
+            else:
+                step = distribution.step
+            bounds = (distribution.low, distribution.high)
+            param_value = float(
+                untransform_one_param(transformed_param[i], param_type, bounds, step)
```

Comment:
The typing is incorrect based on `untransform_one_param`.
```suggestion
                untransform_one_param(np.asarray([transformed_param[i]]), param_type, bounds, step)[0]
```

Reply:
Hmm...I don't think this would be more readable. I thought `float` type was semantically convertible to `np.ndarray` (as 0-dimensional array, and that can also be done explicitly by `np.array(transformed_param[i])`). Also mypy does not complain about this. How about leaving the original line as it is?

Reply:
Actually, I did not know the spec.
Then could you add an inline comment above about this spec?